Creating a restful api

Yosziad opened the pull request.

Yosziad committed:
Creating a restful api

benios commented on server/app.js:
```diff
@@ -0,0 +1,40 @@
+    logger.error("not found"),{
+        "status":404
+    }
+});
```
I don't think there is a need for two functions

benios commented on server/package.json:
```diff
@@ -0,0 +1,26 @@
+    "body-parser": "^1.19.0",
+    "express": "^4.17.1",
+    "morgan": "^1.10.0",
+    "winston": "^3.3.3"
```
you have both motgan and winston that are both logger packages, use only one

benios commented on server/routes/order.js:
```diff
@@ -0,0 +1,17 @@
+const express = require("express");
+const router = express.Router();
+const controller = require("../controllers/order");
```
what will happen if you will use two different controllers?
rename controller to orderController

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,94 @@
+const express = require ("express");
+const bodyParser = require("body-parser");
+const ordersModel = require("../model/order");
```
orderModel

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,94 @@
+    },
+};  
+
+function dateGenerator() {
```
there is no need to save the date in readable format, saving new Date() is the right way
and when you need to show readable date to the user parse it

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,94 @@
+    logger.setLogData(order);
+    logger.info("request recieved at /order");
+
+    if (body.userName == null || body.userName == ""){
```
===

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,94 @@
+
+const createOrder = (req, res) => { 
+    const body = req.body;
+    let error = {};
```
const

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,94 @@
+
+    if (body.userName == null || body.userName == ""){
+        logger.error("username field is empty");
+        error["username"] = "username field is empty";
```
error.username = "username field is empty";

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,94 @@
+    if (body.product == null || body.product == ""){
+        logger.error("product field is empty");
+        error["product"] = "product field is empty";
+    };
```
same as above

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,94 @@
+        logger.error("product field is empty");
+        error["product"] = "product field is empty";
+    };
+    if(Object.keys(error).length !=0){
```
length > 0

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,94 @@
+        logger.error("creating a new order failed"),{
+            "success":false
+        }
+        res.send("Error")
```
return res.send("Error");

benios replied:
It is better if you return the error, like "product field is empty" and not "Error"

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,94 @@
+            "success":false
+        }
+        res.send("Error")
+    }else{
```
remove else

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,94 @@
+    const order = {
+        userName: body.userName,
+        product: body.product,
+        date: dateGenerator(),
```
new Date()

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,94 @@
+    let error = {};
+    const order = {
+        userName: body.userName,
+        product: body.product,
```
the requirement was that a user can order how many products as he want, and not only one per order

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,94 @@
+    const id = Number(req.params.orderId);
+    const order = ordersModel.getOrder(id);
+    logger.setLogData(order);
+    logger.info("order with " + id + " id was fetched");
```
use format string as i showed you
```
`order ${id} was fetched`
```

benios commented on server/model/product.js:
```diff
@@ -0,0 +1,27 @@
+
+const getProducts = () => (products);
+
+const getProductId = (id) => products.find(product => product.id === id);
```
getProductById

benios commented on server/model/product.js:
```diff
@@ -0,0 +1,27 @@
+
+const updateProduct = (product) => {
+    let oldProduct = (element) => element.id === product.id;
+    let index = products.findIndex(oldProduct);
```
use const, and you can just enter the function inside findIndex and delete oldProduct var

benios commented on server/model/product.js:
```diff
@@ -0,0 +1,27 @@
+
+const deleteProduct = (id) => {
+    let Product = (element) => element.id === id;
+    let index = products.findIndex(Product);
```
same as above

benios commented on server/model/user.js:
```diff
@@ -0,0 +1,17 @@
+
+const deleteUser = (id) => {
+    let user = (element) => element.id === id;
+    let index = usersList.findIndex(user);
```
use const, and you can just enter the function inside findIndex and delete use var

benios commented on server/controllers/user.js:
```diff
@@ -0,0 +1,122 @@
+app.use(bodyParser.json());  
+
+
+const idGeneretor = {
```
you are using the same code in 3 places, what will you do id you will have 100 controllers and you will want to change this code?
you need to find a solution that will not cause code duplication

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,155 @@
+	},
+};
+
+function dateGenerator() {
```
same

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,155 @@
+		quantity: req.body.quantity,
+		pictureUrl: req.body.pictureUrl,
+		description: req.body.description,
+		date: dateGenerator(),
```
new Date()

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,155 @@
+	};
+
+	logger.setLogData(product);
+	logger.info("request recieved at /product");
```
non informative message, maybe something like "creating a product"

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,155 @@
+	if (product.price == null || product.price == "") {
+		logger.error("price field is empty");
+		error["price"] = "price field is empty";
+	}
```
===, error.name, error.price

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,155 @@
+	}
+	var priceValidationRegEx = /^([0-9])+$/;
+	var isPriceValid = priceValidationRegEx.test(product.price);
+	if (!isPriceValid) {
```
I don't think regex is a good solution here, you need to check if the data is a number, you can use Number.isInteger()

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,155 @@
+	}
+	if (product.quantity == null || product.quantity == "") {
+		logger.error("quantity field is empty");
+		error["quantity"] = "quantity field is empty";
```
same

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,155 @@
+	}
+	var quantityValidationRegEx = /^([0-9])+$/;
+	var isQuantityValid = quantityValidationRegEx.test(product.quantity);
+	if (!isQuantityValid) {
```
same as price

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,155 @@
+	if (product.description == null || product.description == "") {
+		logger.error("description field is empty");
+		error["description"] = "description field is empty";
+	}
```
same

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,155 @@
+			{
+				success: false,
+			};
+		res.send("Error");
```
> 0 , and I think this is not a good practive to return "Error", just return the specific error when you find it

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,155 @@
+	});
+};
+
+const getProductId = (req, res) => {
```
getProductById

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,94 @@
+
+const getAllOrders = (req, res) => { 
+    const ordersList = ordersModel.getAllOrders();
+    logger.setLogData(ordersList);
+    logger.info("Orders were fetched");
```
I think it will be better to just use info(message, obj), I need to check, don't change it for now

benios commented on server/controllers/user.js:
```diff
@@ -0,0 +1,122 @@
+    logger.setLogData(user);
+    logger.info("request recieved at /user");
+
+    if (user.userName == null || user.userName == ""){
+        logger.error("username field is empty");
+        error["username"] = "username field is empty";
+    };
+    var userNameValidationRegEx = /^([a-z]|[0-9]|-|_)+$/;
```
why not just use regex?

benios commented on server/controllers/user.js:
```diff
@@ -0,0 +1,122 @@
+        logger.error("userName field is incorrect");
+		error["userName"] = "userName field is incorrect";
+	} 
+    if (user.password == null || user.password == ""){
+        logger.error("password field is empty");
+        error["password"] = "password field is empty";
+    };
+	if (user.password < 8) {
+		logger.error("password field is incorrect");
+		error["password"] = "password field is incorrect";
+	}
```
one check should be enough, you dont want to give hackers any info if they try to still a user's password

benios commented on server/controllers/user.js:
```diff
@@ -0,0 +1,122 @@
+		logger.error("password field is incorrect");
+		error["password"] = "password field is incorrect";
+	}
+    if (user.firstName == null || user.firstName == ""){
+        logger.error("firstName field is empty");
+        error["firstName"] = "firstName field is empty";
+    };
+    if (user.lastName == null || user.lastName == ""){
+        logger.error("lastName field is empty");
+        error["lastName"] = "lastName field is empty";
+    };
+    if (user.address == null || user.address == ""){
+        logger.error("address field is empty");
+        error["address"] = "address field is empty";
+    };
+    if (user.age == null || user.age == ""){
+        logger.error("age field is empty");
+        error["age"] = "age field is empty";
+    };
+    var ageValidationRegEx = /^([0-9])+$/;
+	var isAgeValid = ageValidationRegEx.test(user.age);
+	if (!isAgeValid) {
+		logger.error("age field is not a number");
+		error["age"] = "age field is not a number";
+	}
+    if(Object.keys(error).length !=0){
+        logger.error("creating a user failed"),{
+            "success":false
+        }
+        res.send("Error")
+    }else{
+        usersModel.createUser(user);
+        logger.info("user created successfully"),{
+            "success":true 
+        }
+        res.status(201).json({
+            message: "User created",
+            createdUser: user
+        });
+    }
```
same as in other controllers

benios commented:
there are some code standards issue, lets install npm esling, amd use airbnb plugin to make sure that the code will be on standard

benios commented:
you didn't implement the email notification when user create an order

benios commented:
○	שליפת כל ההזמנות של משתמש.

benios commented:
תעשה גם שההזמנות יחזרו בסדר כך שהחדש ביותר יהיה ראשון

Yosziad committed:
rest api update

adding email notification , fixing some faults and addding get orders by username

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,111 @@
+  service: 'gmail',
+  auth: {
+    user: 'ziadeyusef@gmail.com',
+    pass: 'Yos112233',
```
open dummy mail

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,111 @@
+    this.lastId += 1;
+    return this.lastId;
+  },
+};
```
create utils folder with function that will handle id generation

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,111 @@
+const getAllOrders = (req, res) => {
+  const ordersList = orderModel.getAllOrders();
+  logger.info('Orders were fetched', ordersList);
+  res.status(200).json({
```
return

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,111 @@
+  const order = {
+    userName: body.userName,
+    product: body.product,
+    date: new Date(Date.now()).toLocaleString(),
```
date: new Date(), there is no need in saving the text value

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,111 @@
+
+  if (body.userName === null || body.userName === '') {
+    logger.error('username field is empty', order);
+    error.username = 'username field is empty';
```
return res.send('username field is empty');

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,111 @@
+  }
+  if (body.product === null || body.product === '') {
+    logger.error('product field is empty', order);
+    error.product = 'product field is empty';
```
return res.send('product field is empty');

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,111 @@
+  if (Object.keys(error).length > 0) {
+    logger.error('creating a new order failed', order);
+    return res.send('creating a new order failed');
+  }
```
remove this if

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,111 @@
+  const id = Number(req.params.orderId);
+  const order = orderModel.getOrder(id);
+  logger.info(`order with ${id} id was fetched`, order);
+  res.status(200).json({
```
return

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,111 @@
+  const usersOrders = orderModel.getOrdersByUsername(username);
+  const sortedOrders = usersOrders.sort((a, b) => (Date.parse(b.date) - Date.parse(a.date)));
+  logger.info(`orders of username:${username}has been fetched`, sortedOrders);
+  res.status(200).json({
```
return

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,111 @@
+  const username = req.params.usersOrders;
+  logger.info('fetching users orders');
+  const usersOrders = orderModel.getOrdersByUsername(username);
+  const sortedOrders = usersOrders.sort((a, b) => (Date.parse(b.date) - Date.parse(a.date)));
```
sort in the model, this is not controller related

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,119 @@
+    return this.lastId;
+  },
+};
+
```
same

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,119 @@
+  logger.info('handling create a product request', product);
+
+  if (product.name === null || product.name === '') {
+    logger.error('name field is empty', product);
```
return res.send('name field is empty');

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,119 @@
+  const isPriceValid = priceValidationRegEx.test(product.price);
+  if (!isPriceValid) {
+    logger.error('price field is not a number', product);
+  }
```
if (!isNan(product.price)) {
    logger.error('price field is not a number', product);
    return res.send('price field is not a number');
  }

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,119 @@
+  const isQuantityValid = quantityValidationRegEx.test(product.quantity);
+  if (!isQuantityValid) {
+    logger.error('Quantity field is not a number', product);
+  }
```
same as above, use isNan

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,119 @@
+    logger.error('Quantity field is not a number', product);
+  }
+  if (product.pictureUrl === null || product.pictureUrl === '') {
+    logger.error('pictureUrl field is empty', product);
```
return res

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,119 @@
+    logger.error('pictureUrl field is empty', product);
+  }
+  if (product.description === null || product.description === '') {
+    logger.error('description field is empty', product);
```
return res

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,119 @@
+  if (Object.keys(error).length > 0) {
+    logger.error('adding product failed', product);
+    return res.send('adding product failed');
+  }
```
remove if

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,119 @@
+  const product = productModel.getProductId(id);
+
+  logger.info(`Handling GET requests to /product/${id}`, product);
+  res.status(200).json({
```
return

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,119 @@
+  };
+  productModel.updateProduct(product);
+  logger.info('Updated product!', product);
+  res.status(200).json({
```
return

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,119 @@
+    quantity: req.body.quantity,
+    pictureUrl: req.body.pictureUrl,
+    description: req.body.description,
+    date: new Date(Date.now()).toLocaleString(),
```
date: new Date(),

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,119 @@
+  const id = Number(req.params.productId);
+  const deletedProduct = productModel.deleteProduct(id);
+  logger.info('Deleted product!', deletedProduct);
+  res.status(200).json({
```
return

benios commented on server/controllers/user.js:
```diff
@@ -0,0 +1,92 @@
+  const user = userModel.getUser(id);
+  logger.setLogData(user);
+  logger.info('You passed a user ID');
+  res.status(200).json({
```
return

benios commented on server/controllers/user.js:
```diff
@@ -0,0 +1,92 @@
+
+  const userNameValidationRegEx = /^([a-z]|[0-9]|-|_)+$/;
+  const isUserNameValid = userNameValidationRegEx.test(user.userName);
+  if (!isUserNameValid) {
```
same validation issues

benios commented on server/controllers/user.js:
```diff
@@ -0,0 +1,92 @@
+  const id = Number(req.params.userId);
+  const deletedUser = userModel.deleteUser(id);
+  logger.info('User deleted!', deletedUser);
+  res.status(200).json({
```
return

benios commented on server/model/product.js:
```diff
@@ -0,0 +1,27 @@
+const updateProduct = (product) => {
+  const oldProduct = (element) => element.id === product.id;
+  const index = products.findIndex(oldProduct);
+  products[index] = product;
```
const productsndex = products.findIndex((element) => element.id === product.id);
  products[index] = product;

benios commented on server/model/product.js:
```diff
@@ -0,0 +1,27 @@
+
+const deleteProduct = (id) => {
+  const Product = (element) => element.id === id;
+  const index = products.findIndex(Product);
```
name should start in small letter,
const productIndex = products.findIndex((element) => element.id === product.id);

benios commented on server/model/user.js:
```diff
@@ -0,0 +1,17 @@
+const getUser = (id) => usersList.find((user) => user.id === id);
+
+const deleteUser = (id) => {
+  const user = (element) => element.id === id;
+  const index = usersList.findIndex(user);
```
const userIndex = products.findIndex((element) => element.id === id);

Yosziad committed:
update 2

Yosziad committed:
update 3

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,99 @@
+
+  logger.info('creating an order', order);
+
+  if (body.userName === null || body.userName === '') {
```
You can do if (!body.userName). ! on empty string is True. Also you didn't check if body.userName is Undefined

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,99 @@
+    logger.error('username field is empty', order);
+    return res.send('username field is empty');
+  }
+  if (body.product === null || body.product === '') {
```
same as body.userName

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,99 @@
+app.use(bodyParser.urlencoded({ extended: true }));
+app.use(bodyParser.json());
+
+const transporter = nodemailer.createTransport({
```
Open new File with Email services that export you send email function that you will use here. that way you can send mail from anywhere without define it over again.

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,99 @@
+  logger.info('Orders were created', order);
+  const strOrder = JSON.stringify(order);
+
+  const mailOptions = {
```
Do this in the email service function

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,99 @@
+
+const getOrder = (req, res) => {
+  const id = Number(req.params.orderId);
+  const order = orderModel.getOrder(id);
```
what if the id didn't exist or if the req.params.orderId isn't a number?

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,99 @@
+const getOrdersByUsername = (req, res) => {
+  const username = req.params.usersOrders;
+  logger.info('fetching users orders');
+  const usersOrders = orderModel.getOrdersByUsername(username);
```
Same as before, what if this username isn't exist

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,109 @@
+
+  logger.info('handling create a product request', product);
+
+  if (product.name === null || product.name === '') {
```
You can use !

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,109 @@
+    logger.error('Quantity field is not a number', product);
+    return res.send('Quantity field is not a number');
+  }
+  if (product.pictureUrl === null || product.pictureUrl === '') {
```
you can use !

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,109 @@
+
+const getProductById = (req, res) => {
+  const id = Number(req.params.productId);
+  const product = productModel.getProductId(id);
```
What if this id isn't exist. you need to send another res status to the client

benios commented on server/controllers/order.js:
```diff
@@ -0,0 +1,99 @@
+    userName: body.userName,
+    product: body.product,
+    date: new Date(),
+    id: utils.idGeneretor.generateId(),
```
The id is usually generated in the db, because you didn't have one I think this  utils.idGeneretor.generateId() need to be in the model layer. the controller didn't need to determine the id.

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,109 @@
+app.use(bodyParser.json());
+
+const createProduct = (req, res) => {
+  const id = utils.idGeneretor.generateId();
```
The id is usually generated in the db, because you didn't have one I think this  utils.idGeneretor.generateId() need to be in the model layer. the controller didn't need to determine the id.

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,109 @@
+    date: new Date(),
+    id,
+  };
+  productModel.updateProduct(product);
```
Check what you get, what if this id didn't exist and you get null from the model?

benios commented on server/controllers/product.js:
```diff
@@ -0,0 +1,109 @@
+
+const deleteProduct = (req, res) => {
+  const id = Number(req.params.productId);
+  const deletedProduct = productModel.deleteProduct(id);
```
same here, check if the product really deleted or not been exist.

benios commented on server/controllers/user.js:
```diff
@@ -0,0 +1,84 @@
+
+const getUser = (req, res) => {
+  const id = Number(req.params.userId);
+  const user = userModel.getUser(id);
```
Same here, do res with error when the user not found

benios commented on server/controllers/user.js:
```diff
@@ -0,0 +1,84 @@
+    lastName: req.body.lastName,
+    address: req.body.address,
+    age: req.body.age,
+    id: utils.idGeneretor.generateId(),
```
utils.idGeneretor.generateId() in the model

benios commented on server/controllers/user.js:
```diff
@@ -0,0 +1,84 @@
+  const userNameValidationRegEx = /^([a-z]|[0-9]|-|_)+$/;
+  const isUserNameValid = userNameValidationRegEx.test(user.userName);
+  if (!isUserNameValid) {
+    logger.error('Either userName or password field is incorrect');
```
send res

benios commented on server/controllers/user.js:
```diff
@@ -0,0 +1,84 @@
+
+const deleteUser = (req, res) => {
+  const id = Number(req.params.userId);
+  const deletedUser = userModel.deleteUser(id);
```
Same here, do res with error when the user not found

benios commented on server/model/order.js:
```diff
@@ -0,0 +1,20 @@
+
+const getOrder = (id) => ordersList.find((order) => order.id === id);
+
+const createOrder = (order) => ordersList.push(order);
```
generate the id here and then push it with new id

benios commented on server/model/order.js:
```diff
@@ -0,0 +1,20 @@
+
+const getOrdersByUsername = (username) => {
+  const usersOrders = ordersList.filter((order) => order.userName === username);
+  const sortedOrders = usersOrders.sort((a, b) => (Date.parse(b.date) - Date.parse(a.date)));
```
Please check if it works.

benios commented on server/model/product.js:
```diff
@@ -0,0 +1,25 @@
+const createProduct = (product) => products.push(product);
+
+const updateProduct = (product) => {
+  const index = products.findIndex((element) => element.id === product.id);
```
what if this index not found

benios commented on server/model/user.js:
```diff
@@ -0,0 +1,16 @@
+const getUser = (id) => usersList.find((user) => user.id === id);
+
+const deleteUser = (id) => {
+  const index = usersList.findIndex((element) => element.id === id);
```
what if this index not found

benios commented on server/package.json:
```diff
@@ -0,0 +1,32 @@
+  "description": "",
+  "main": "app.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
```
add "start" script and use npm start for starting your app

beniostr replied:
"start": "nodemone app.js"

benios commented on server/routes/order.js:
```diff
@@ -0,0 +1,14 @@
+
+router.get('/orders/:orderId', controller.getOrder);
+
+router.get('/usersOrders/:usersOrders', controller.getOrdersByUsername);
```
instead of  \:usersOrders call it  \:username or \:user

beniostr commented on server/controllers/product.js:
```diff
@@ -0,0 +1,109 @@
+    logger.error('name field is empty', product);
+    return res.send('name field is empty');
+  }
+  if (Number.isNaN(product.price)) {
```
typeof price == 'number' && price >= 0

Yosziad committed:
update 4

Yosziad committed:
update 5

Yosziad committed:
update 6

Yosziad committed:
update 7

beniostr commented on server/services/Email_services.js:
```diff
@@ -0,0 +1,37 @@
+  },
+});
+
+const SERVICE_EMAIL_ADDRESS = {
+  from: 'kingmonkey409@gmail.com',
+  to: 'ziadeyusef@gmail.com',
+};
```
Object property is not const as you can change that at any time,
const SERVICE_EMAIL_ADDRESS = 'kingmonkey409@gmail.com';
const ADMIN_EMAIL_ADDRESS = 'ziadeyusef@gmail.com';

beniostr commented on server/services/Email_services.js:
```diff
@@ -0,0 +1,37 @@
+    text: `${text} :${details}`,
+  };
+
+  transporter.sendMail(mailOptions, (error2, info) => {
```
change function var name from error2 to error

kfir-payne commented on server/controllers/order.js:
```diff
@@ -0,0 +1,86 @@
+
+const app = express();
+
+const Logger = require('../services/logger_services');
```
To keep the code clean and readable all the require/import statements will be in one place, at the top of the file.

relevant to all the appearances of this mistake.

Yosziad committed:
update 8

kfir-payne commented on server/package.json:
```diff
@@ -0,0 +1,33 @@
+  "devDependencies": {
+    "eslint": "^7.10.0",
+    "eslint-config-airbnb-base": "^14.2.0",
+    "eslint-plugin-import": "^2.22.1"
```
add nodemon to devDependencies so it will be installed in someone else's machine, you probably installed it locally in your machine.

Yosziad committed:
update 9

kfir-payne commented on server/controllers/product.js:
```diff
@@ -0,0 +1,120 @@
+  if (!product.description) {
+    logger.error('description field is empty', product);
+    return res.send('description field is empty');
+  }
```
I suggest to move all validation logic to a new function and handle all the error cases there.

relevant to all the appearances of validation checks.

kfir-payne commented on server/controllers/order.js:
```diff
@@ -0,0 +1,86 @@
+    return res.send('product field is empty');
+  }
+  orderModel.createOrder(order);
+  logger.info('Orders were created', order);
+  const strOrder = JSON.stringify(order);
+  const subject = 'New order';
+  const text = 'A new order has been created';
+  emailNotification.emailNotification(strOrder, subject, text);
```
Would move all of this to a different function responsible of email notification

kfir-payne commented on server/controllers/product.js:
```diff
@@ -0,0 +1,120 @@
+    description: req.body.description,
+    date: new Date(),
+    id,
+  };
```
Redundant line of code, can be modified like this:
  ```
const product = {
    name: req.body.name,
    price: req.body.price,
    quantity: req.body.quantity,
    pictureUrl: req.body.pictureUrl,
    description: req.body.description,
    date: new Date(),
    id: Number(req.params.productId),
  };
```

kfir-payne commented on server/model/product.js:
```diff
@@ -0,0 +1,43 @@
+
+const createProduct = (product) => {
+  const id = utils.idGeneretor.generateId();
+  products.push(...product, id);
```
Missing {} in the push

kfir-payne commented on server/model/user.js:
```diff
@@ -0,0 +1,20 @@
+
+const createUser = (user) => {
+  const id = utils.idGeneretor.generateId();
+  usersList.push(...user, id);
```
Missing {} in the push

kfir-payne commented on server/services/Email_services.js:
```diff
@@ -0,0 +1,35 @@
+  service: 'GMAIL',
+  auth: {
+    user: SERVICE_EMAIL_ADDRESS,
+    pass: 'Yos112233',
```
Consider moving 'service' & 'pass' to const config vars like you did with SERVICE_EMAIL_ADDRESS & ADMIN_EMAIL_ADDRESS

kfir-payne commented on server/utils/utils.js:
```diff
@@ -0,0 +1,9 @@
+const idGeneretor = {
```
Typo in Generetor should be Generator, dumb Beni

kfir-payne commented on server/services/logger_services.js:
```diff
@@ -0,0 +1,54 @@
+    });
+  }
+}
+module.exports = LoggerService;
```
Look, there's an error in this line, "No new line at end of file"

kfir-payne commented on server/controllers/product.js:
```diff
@@ -0,0 +1,120 @@
+  }
+  productModel.createProduct(product);
+  logger.info('added product seccessfully', product);
+  return res.send('added product seccessfully');
```
Typo in successfully

kfir-payne commented on server/app.js:
```diff
@@ -0,0 +1,28 @@
+  error.status = 404;
+  res.status(error.status || 500);
+  logger.error('not found');
+});
```
You are not returning a response to the client, try sending a request to a non-existing route and see what happens.

kfir-payne commented on server/controllers/user.js:
```diff
@@ -0,0 +1,83 @@
+  }
+  if (!user.password || (user.password).length < 8) {
+    logger.error('Either userName or password field is incorrect');
+    return res.send('Either userName or password field is incorrect');
```
Try sending a more indicative error message, so the user can know what he did wrong.

kfir-payne commented on server/controllers/user.js:
```diff
@@ -0,0 +1,83 @@
+app.use(bodyParser.urlencoded({ extended: true }));
+app.use(bodyParser.json());
+
+const getUser = (req, res) => {
+  const id = Number(req.params.userId);
+  const user = userModel.getUser(id);
+  logger.setLogData(user);
+  logger.info('You passed a user ID');
+  return res.status(200).json({
+    message: 'You passed a user ID',
+    user,
+  });
+};
```
In general, (not specifically in this function but also here) you should handle all kind of errors and return it to the client.
for instance, in this function what happens if the userId provided not found?

fix in all relevant places.

Yosziad committed:
update 10

beniostr commented on server/controllers/product.js:
```diff
@@ -0,0 +1,123 @@
+  if (!product.name) {
+    throw new Error('name field is empty');
+  }
+  if (!(typeof (product.price) === 'number') || !(product.price >= 0)) {
```
you can make the if more simple to read:
if (typeof (product.price) !== 'number' || product.price < 0)

beniostr commented on server/controllers/product.js:
```diff
@@ -0,0 +1,123 @@
+  if (!(typeof (product.price) === 'number') || !(product.price >= 0)) {
+    throw new Error('price field is not a number');
+  }
+  if (!(typeof (product.quantity) === 'number') || !(product.quantity >= 0)) {
```
you can make the if more simple to read

beniostr commented on server/controllers/product.js:
```diff
@@ -0,0 +1,123 @@
+
+const deleteProduct = (req, res) => {
+  const id = Number(req.params.productId);
+  const deletedProduct = productModel.deleteProduct(id);
```
variable with bool type should be named as questions
const didProductDeleted = productModel.deleteProduct(id)

beniostr commented on server/controllers/product.js:
```diff
@@ -0,0 +1,123 @@
+      message: 'product id does not exist, deleting the product failed',
+    });
+  }
+  logger.info('Deleted product!', deletedProduct);
```
what is being logged here?
'Deleted product! true' ?

beniostr commented on server/controllers/user.js:
```diff
@@ -0,0 +1,98 @@
+  if (!user.address) {
+    throw new Error('address field is empty');
+  }
+  if (!!(typeof (user.age) === 'number') || !(user.age >= 0)) {
```
!!(typeof (user.age) === 'number') means ==> (typeof (user.age) === 'number'), so the check is broken?
and simplify the if

beniostr commented on server/model/order.js:
```diff
@@ -0,0 +1,33 @@
+
+const getOrder = (id) => {
+  if (ordersList.some((e) => e.id === id)) {
+    return ordersList.find((order) => order.id === id);
+  }
+  return false;
```
returning two different types of values in a function is not a good practice,
if you want falsy value as we talked about you can use null, as null is typeof object

beniostr replied:
you can just do this for this function
const getOrder = (id) => {
  return ordersList.find((order) => order.id === id);
}
as find function will return undefined when there is no item found, and undefined is falsy value

Yosziad committed:
update 11

beniostr commented on server/model/order.js:
```diff
@@ -0,0 +1,33 @@
+};
+
+const getOrdersByUsername = (username) => {
+  if (ordersList.some((e) => e.userName === username)) {
+    const usersOrders = ordersList.filter((order) => order.userName === username);
+    usersOrders.sort((a, b) => (b.date - a.date));
+    return usersOrders;
+  }
```
There is no need for finding orders of user with some function
If you use filter function and there are no items return the empty list.
this is not an error.
if the user doesn't exist in the user list this should be an error

beniostr commented on server/model/product.js:
```diff
@@ -0,0 +1,43 @@
+};
+
+const updateProduct = (product) => {
+  if (products.some((e) => e.id === product.id)) {
```
returning the updated item is better for later use

beniostr commented on server/model/user.js:
```diff
@@ -0,0 +1,31 @@
+  usersList.push({ ...user, id });
+};
+const getUser = (id) => {
+  if (usersList.some((e) => e.id === id)) {
+    return usersList.find((user) => user.id === id);
+  }
+  return false;
+};
```
not the same type returned

Yosziad committed:
update 12
